[DAG] Allow AssertZExt to scalarize.

davemgreen opened the pull request:
With range and undef metadata on a call we can have vector AssertZExt generated on a target with no vector operations. The AssertZExt needs to scalarize to a normal `AssertZext tin, ValueType`. I have added AssertSext too, although I do not have a test case.

Fixes #110374

arsenm commented on llvm/test/CodeGen/ARM/scalarize-assert-zext.ll:
```diff
@@ -0,0 +1,28 @@
+  store <4 x i8> %conv.i.i.i71, ptr %0, align 1
+  ret void
+}
+
```
Test where the assert is useful after legalization would be nice

arsenm commented on llvm/lib/CodeGen/SelectionDAG/LegalizeVectorTypes.cpp:
```diff
@@ -58,7 +58,11 @@ void DAGTypeLegalizer::ScalarizeVectorResult(SDNode *N, unsigned ResNo) {
+  case ISD::AssertZext:
+  case ISD::AssertSext:
+  case ISD::FPOWI:
+    R = ScalarizeVecRes_ExpOp(N);
```
Function name doesnt' quite match anymore

davemgreen committed:
[DAG] Allow AssertZExt to scalarize.

With range and undef metadata on a call we can have vector AssertZExt generated
on a target with no vector operations. The AssertZExt needs to scalarize to a
normal `AssertZext tin, ValueType`. I have added AssertSext too, although I do
not have a test case.

glandium commented:
Would it make sense to push to the 19.x branch and spin a 19.1.7?

glandium commented:
Cc: @tru

nikic commented:
/cherry-pick ab9a80a3ad78f611fd06cd6f7215bd828809310c
